eth2/utils/ssz_types/src/bitfield.rs

```diff
@@ -0,0 +1,1114 @@
+    pub fn from_bytes(bytes: Vec<u8>) -> Option<Self> {
+        let mut initial_bitfield: Bitfield<BitList<N>> = {
+            let num_bits = bytes.len() * 8;
+            Bitfield::from_raw_bytes(bytes, num_bits)
+                .expect("Must have adequate bytes for bit count.")
+        };
+
+        let len = initial_bitfield.highest_set_bit()?;
+
+        if len <= Self::max_len() {
+            initial_bitfield
+                .set(len, false)
+                .expect("Bit has been confirmed to exist");
+
+            let mut bytes = initial_bitfield.into_raw_bytes();
+
+            if bytes_for_bit_len(len) < bytes.len() && bytes != [0] {
+                bytes.remove(0);
+            }
+
+            Self::from_raw_bytes(bytes, len)
+        } else {
+            None
+        }
+    }
+}
+
+impl<N: Unsigned + Clone> Bitfield<BitVector<N>> {
+    /// Instantiate a new `Bitfield` with a fixed-length of `N` bits.
+    ///
+    /// All bits are initialized to `false`.
+    pub fn new() -> Self {
+        Self {
+            bytes: vec![0; bytes_for_bit_len(Self::capacity())],
+            len: Self::capacity(),
+            _phantom: PhantomData,
+        }
+    }
+
+    /// Returns `N`, the number of bits in `Self`.
+    pub fn capacity() -> usize {
+        N::to_usize()
+    }
+
+    /// Consumes `self`, returning a serialized representation.
+    ///
+    /// The output is faithful to the SSZ encoding of `self`.
+    ///
+    /// ## Example
+    /// ```
+    /// use ssz_types::{Bitfield, typenum};
+    ///
+    /// type BitVector = Bitfield<ssz_types::BitVector<typenum::U4>>;
+    ///
+    /// assert_eq!(BitVector::new().into_bytes(), vec![0b0000_0000]);
+    /// ```
+    pub fn into_bytes(self) -> Vec<u8> {
+        self.into_raw_bytes()
+    }
+
+    /// Instantiates a new instance from `bytes`. Consumes the same format that `self.into_bytes()`
+    /// produces (SSZ).
+    ///
+    /// Returns `None` if `bytes` are not a valid encoding.
+    pub fn from_bytes(bytes: Vec<u8>) -> Option<Self> {
+        Self::from_raw_bytes(bytes, Self::capacity())
+    }
+}
+
+impl<N: Unsigned + Clone> Default for Bitfield<BitVector<N>> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<T: BitfieldBehaviour> Bitfield<T> {
+    /// Sets the `i`'th bit to `value`.
+    ///
+    /// Returns `None` if `i` is out-of-bounds of `self`.
+    pub fn set(&mut self, i: usize, value: bool) -> Option<()> {
+        if i < self.len {
+            let byte = {
+                let num_bytes = self.bytes.len();
+                let offset = i / 8;
+                self.bytes
+                    .get_mut(num_bytes - offset - 1)
+                    .expect("Cannot be OOB if less than self.len")
+            };
+
+            if value {
+                *byte |= 1 << (i % 8)
+            } else {
+                *byte &= !(1 << (i % 8))
+            }
+
+            Some(())
+        } else {
+            None
```

Comment:
I quite liked it when these functions (`get`/`set`) returned `Result`, because then it was straight-forward to convert the errors in `state_processing`.

Reply:
i.e.

```rust
state.justification_bits.set(0, true)?;
```

rather than

```
        state
            .justification_bits
            .set(0, true)
            .ok_or(ssz_types::Error::OutOfBounds {
                i: 1,
                len: state.justification_bits.len(),
            })?;
```

Reply:
Yeah I preferred `Result` vs `Option` as well -- I just thought it was a little off compared to `Vec.get(..)`, etc.

If we both prefer `Result`, then lets do that. #listentoyourheart

Reply:
Good point about `Vec`, but I think we have a strong case for the `Result`, and it's always easier to throw information away (`.ok()`) than to conjure it from thin air